[Clang] Fixes of builtin definitions after PR #68324.

michele-scandale opened the pull request:
This commit addresses few differences between the `Builtins.def` file before the refactoring done in PR #68324 and the currently generated `Builtins.inc` file.

michele-scandale commented:
There are still some differences that I haven't addressed here as they might be ok to have:
* `__builtin_rint{,f,l}` were defined with the `const` attribute, now they use the `const when FP exceptions are ignored` -- this seems due to the definition of these has been merged with the corresponding library builtins `rint{,f,l}`.
* `__builtin_{wcschr,wcscmp,wcslen,wcsncmp}` and corresponding library builtins now are marked `pure`
* `__addressof` was language builtin, now is a library builtin (like `addressof`)
* `finite{,f,l}` were defined with `GNU_LANG` property, now they are defined with the `ALL_GNU_LANGUAGES` property
* `__builtin_va_{copy,end}` are now marked with the property that they are a `libc/libm function with a '__builtin_' prefix added`
* a bunch of C library builtins are now marked `nothrow` (e.g. `abort`, `alloca`, `bcopy`, `bzero`, `index`, `fprintf`, `memchr`, `memcpy`, `memmove`, `memset`, `rindex`, `snprintf`, `sprintf`, `stpcpy` `stpncpy`, `strcat`, `strchr`, `strcmp`, `strcpy`, `strdup`, `strlen`, `strncat`, `strncmp`, `strncpy`, `strpbrk`, `strrchr`, `strspn`, `strstr`, `vfprintf`, `vprintf`, `vsnprintf`, `vsprintf`, `wcschr`, `wcscmp`, `wcslen`, `wcsncmp`, `wmemchr`, `wmemcmp`, `wmemcpy`, `wmemmove`
* `snprintf`, `sprintf`, `vfprintf`, `vprintf`, `vsnprintf`, `vsprintf` have now pointer arguments marked `restrict`.

Any comment?

AaronBallman commented:
> There are still some differences that I haven't addressed here as they might be ok to have:
> 
>     * `__builtin_rint{,f,l}` were defined with the `const` attribute, now they use the `const when FP exceptions are ignored` -- this seems due to the definition of these has been merged with the corresponding library builtins `rint{,f,l}`.
> 
>     * `__builtin_{wcschr,wcscmp,wcslen,wcsncmp}` and corresponding library builtins now are marked `pure`
> 
>     * `__addressof` was language builtin, now is a library builtin (like `addressof`)
> 
>     * `finite{,f,l}` were defined with `GNU_LANG` property, now they are defined with the `ALL_GNU_LANGUAGES` property
> 
>     * `__builtin_va_{copy,end}` are now marked with the property that they are a `libc/libm function with a '__builtin_' prefix added`
> 
>     * a bunch of C library builtins are now marked `nothrow` (e.g. `abort`, `alloca`, `bcopy`, `bzero`, `index`, `fprintf`, `memchr`, `memcpy`, `memmove`, `memset`, `rindex`, `snprintf`, `sprintf`, `stpcpy` `stpncpy`, `strcat`, `strchr`, `strcmp`, `strcpy`, `strdup`, `strlen`, `strncat`, `strncmp`, `strncpy`, `strpbrk`, `strrchr`, `strspn`, `strstr`, `vfprintf`, `vprintf`, `vsnprintf`, `vsprintf`, `wcschr`, `wcscmp`, `wcslen`, `wcsncmp`, `wmemchr`, `wmemcmp`, `wmemcpy`, `wmemmove`
> 
>     * `snprintf`, `sprintf`, `vfprintf`, `vprintf`, `vsnprintf`, `vsprintf` have now pointer arguments marked `restrict`.
> 
> 
> Any comment?

Thank you for catching these mistakes! I'm of two minds: some of these changes seem valid and correct, but I don't think they were intended. For example, `printf` family functions now having pointers marked `restrict` maps to the standard signature for them and seems like a good change. But a change like with `__addressof` may be more questionable. Perhaps we should restore the old behavior in all cases, then re-land the good changes in follow-ups. That's more effort now, but it's easier for people to track down why a change was made in a few years. WDYT?

AaronBallman reviewed (approved):
LGTM, these fixes all look correct to me. The only thing that seemed somewhat off was with the `sync` functions taking a pointer to a `volatile` (that's not documented in GCC's docs, but I think the signature does make sense). Thank you for the fixes!

michele-scandale commented:
> > There are still some differences that I haven't addressed here as they might be ok to have:
> > ```
> > * `__builtin_rint{,f,l}` were defined with the `const` attribute, now they use the `const when FP exceptions are ignored` -- this seems due to the definition of these has been merged with the corresponding library builtins `rint{,f,l}`.
> > 
> > * `__builtin_{wcschr,wcscmp,wcslen,wcsncmp}` and corresponding library builtins now are marked `pure`
> > 
> > * `__addressof` was language builtin, now is a library builtin (like `addressof`)
> > 
> > * `finite{,f,l}` were defined with `GNU_LANG` property, now they are defined with the `ALL_GNU_LANGUAGES` property
> > 
> > * `__builtin_va_{copy,end}` are now marked with the property that they are a `libc/libm function with a '__builtin_' prefix added`
> > 
> > * a bunch of C library builtins are now marked `nothrow` (e.g. `abort`, `alloca`, `bcopy`, `bzero`, `index`, `fprintf`, `memchr`, `memcpy`, `memmove`, `memset`, `rindex`, `snprintf`, `sprintf`, `stpcpy` `stpncpy`, `strcat`, `strchr`, `strcmp`, `strcpy`, `strdup`, `strlen`, `strncat`, `strncmp`, `strncpy`, `strpbrk`, `strrchr`, `strspn`, `strstr`, `vfprintf`, `vprintf`, `vsnprintf`, `vsprintf`, `wcschr`, `wcscmp`, `wcslen`, `wcsncmp`, `wmemchr`, `wmemcmp`, `wmemcpy`, `wmemmove`
> > 
> > * `snprintf`, `sprintf`, `vfprintf`, `vprintf`, `vsnprintf`, `vsprintf` have now pointer arguments marked `restrict`.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Any comment?
> 
> Thank you for catching these mistakes! I'm of two minds: some of these changes seem valid and correct, but I don't think they were intended. For example, `printf` family functions now having pointers marked `restrict` maps to the standard signature for them and seems like a good change. But a change like with `__addressof` may be more questionable. Perhaps we should restore the old behavior in all cases, then re-land the good changes in follow-ups. That's more effort now, but it's easier for people to track down why a change was made in a few years. WDYT?

I can do that.

michele-scandale commented:
Added the change to restore `__addressof` to the prior definition. To do that I have to introduce the `FunctionWithoutBuiltinPrefix` attribute (expanded to `f`) -- this is implicitly added for builtins deriving from `LibBuiltin`, but for this case we are expecting to declare `__addressof` as a language builtin.

michele-scandale commented:
@AaronBallman, ping.

michele-scandale commented:
@AaronBallman, ping.

AaronBallman reviewed (approved):
LGTM, thank you for the fix!

michele-scandale committed:
[Clang] Fixes of builtins definitions after PR #68324.

This commit addresses few differences between the `Builtins.def` file
before the refactoring done in PR #68324 and the currently generated
`Builtins.inc` file.
